src/toxcore/dht/dht_node.rs

```diff
@@ -64,43 +81,68 @@ impl DhtNode {
             (None, Some(clock_now()))
         };
 
+        let assoc4 = SockAndTime {
+            saddr: saddr_v4,
+            last_resp_time: last_resp_time_v4,
+            last_ping_req_time: None,
+            ret_saddr: None,
+            ret_last_resp_time: None,
+        };
+
+        let assoc6 = SockAndTime {
+            saddr: saddr_v6,
+            last_resp_time: last_resp_time_v6,
+            last_ping_req_time: None,
+            ret_saddr: None,
+            ret_last_resp_time: None,
+        };
+
         DhtNode {
             pk: pn.pk,
-            saddr_v4,
-            saddr_v6,
-            last_resp_time_v4,
-            last_resp_time_v6,
-            last_ping_req_time: None,
+            assoc4,
+            assoc6,
         }
     }
 
     /// Check if the node is considered bad i.e. it does not answer both on IPv4
     /// and IPv6 addresses for `BAD_NODE_TIMEOUT` seconds.
     pub fn is_bad(&self) -> bool {
-        self.last_resp_time_v4.map_or(true, |time| clock_elapsed(time) > Duration::from_secs(BAD_NODE_TIMEOUT)) &&
-            self.last_resp_time_v6.map_or(true, |time| clock_elapsed(time) > Duration::from_secs(BAD_NODE_TIMEOUT))
+        self.is_ipv4_bad() && self.is_ipv6_bad()
+    }
+
+    /// Check if the IPv4 address is considered bad i.e. it does not answer on IPv4
+    /// addresses for `BAD_NODE_TIMEOUT` seconds.
+    fn is_ipv4_bad(&self) -> bool {
+        self.assoc4.last_resp_time.map_or(true, |time| clock_elapsed(time) > Duration::from_secs(BAD_NODE_TIMEOUT))
+    }
+
+    /// Check if the IPv6 address is considered bad i.e. it does not answer on IPv6
+    /// addresses for `BAD_NODE_TIMEOUT` seconds.
+    fn is_ipv6_bad(&self) -> bool {
+        self.assoc6.last_resp_time.map_or(true, |time| clock_elapsed(time) > Duration::from_secs(BAD_NODE_TIMEOUT))
     }
 
     /// Check if the node is considered discarded i.e. it does not answer both
     /// on IPv4 and IPv6 addresses for `KILL_NODE_TIMEOUT` seconds.
     pub fn is_discarded(&self) -> bool {
-        self.last_resp_time_v4.map_or(true, |time| clock_elapsed(time) > Duration::from_secs(KILL_NODE_TIMEOUT)) &&
-            self.last_resp_time_v6.map_or(true, |time| clock_elapsed(time) > Duration::from_secs(KILL_NODE_TIMEOUT))
+        self.assoc4.last_resp_time.map_or(true, |time| clock_elapsed(time) > Duration::from_secs(KILL_NODE_TIMEOUT)) &&
```

Comment:
add method SockAndTime::is_discarded and use it here

Reply:
done